numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1770,23 +1783,72 @@ array_iter(PyArrayObject *arr)
     return PySeqIter_New((PyObject *)arr);
 }
 
-static PyObject *
-array_alloc(PyTypeObject *type, Py_ssize_t NPY_UNUSED(nitems))
+
+/*
+ * Impelement traversal for cyclic garbage collection. There are two main
+ * cases. Arrays that do not own their own data, must VISIT their base.
+ * Arrays which hold the actual data own the references and have to visit
+ * all items.
+ */
+static int
+array_traverse(PyArrayObject *self, visitproc visit, void *arg)
 {
-    /* nitems will always be 0 */
-    PyObject *obj = PyObject_Malloc(type->tp_basicsize);
-    PyObject_Init(obj, type);
-    return obj;
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    /*
+     * make sure we don't traverse the array before it is fully initialized,
+     * or if it doesn't have any objects
+     */
+    if (!descr) {
+        return 0;
+    }
+    Py_VISIT(((PyArrayObject_fields *)self)->base);
+
+    if (!PyDataType_REFCHK(descr)) {
+        return 0;
+    }
+    if (!PyArray_CHKFLAGS(self, NPY_ARRAY_OWNDATA)) {
+        /* if the array doesn't own its data, visiting fa->base is sufficient */
+        return 0;
+    }
+    /* NOTE: In the future, user dtypes could in principle be visited */
```

Comment:
Shouldn't the dtype be visited anyway, since it has metadata that could refer back to the array? No need to implement the visiting in the dtype object yet, but I think we should call visit here to play it safe for when we fix that.

Reply:
It is pretty esoteric, but yes. We should visit the descriptor maybe. That would mean we cannot untrack arrays though, since any array has a dtype. (at least unless we also check that the dtype has no metadata). I think I meant the comment as visiting a user-dtypes items (which is much further off).

Not sure what to best do. Could just not do the untracking, or do it anyway, but visit. Either way, this would fix the vast majority of use-cases.